### PR TITLE
Add CTC API

### PR DESCRIPTION
**Pull request by vineelpratap**

### CODE REVIEW

**Comment by awni** [comment]
Can you do a PR with just your changes? For some reason it shows mostly my changes in your PR...

File: gtn/functions.cpp [hunk 1]
```diff
@@ -110,4 +110,27 @@ DISPATCH1(forwardScore)
 DISPATCH1(viterbiScore)
 DISPATCH1(viterbiPath)
 
+Graph ctcLoss(
```
Comment (awni) [inline gtn/functions.cpp:113]: I think this should live in a new file called `criteria` and maybe even in a sub namespace like `gtn::criteria::ctc`. What do you think about that?

Comment (vineelpratap) [reply gtn/functions.cpp:113]: sure... that sounds good

**Comment by vineelpratap** [comment]
- Address comments
- Update python API
- Add docs  
Screen Shot 2021-12-25 at 1 41 06 PM

Screen Shot 2021-12-25 at 1 41 15 PM

Followups (in a separate PR)
- benchmarks

**Comment by vineelpratap** [comment]
I'm merging this PR (although there are some merge conflicts). Will send a follow up PR soon to fix those conflicts.

File: gtn/criterions.cpp [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    int label = l % 2 ? target[idx] : blankIdx;
+    gLabel.addArc(l, l, label);
+    if (l > 0) {
+      gLabel.addArc(l - 1, l, label);
+    }
+    if (l % 2 and l > 1 and label != target[idx - 1]) {
+      gLabel.addArc(l - 2, l, label);
+    }
+  }
+  // gLabel.arcSort(false);
```
Comment (awni) [inline gtn/criterions.cpp:38]: Curious, why did you comment this?

Comment (vineelpratap) [reply gtn/criterions.cpp:38]: I missed commenting it out... I guess I'll enable this only for cpu backend. 

Will send a PR.

**Pull request merged**